satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,307 @@
+parallax correction as a modifier in Satpy.
+"""
+
+import inspect
+import warnings
+from datetime import datetime
+
+import numpy as np
+import xarray as xr
+from pyorbital.orbital import A as EARTH_RADIUS
+from pyorbital.orbital import get_observer_look
+from pyresample.geometry import SwathDefinition
+from pyresample.kd_tree import resample_nearest
+
+from satpy import Scene
+from satpy.modifiers import ModifierBase
+from satpy.utils import get_satpos, lonlat2xyz, xyz2lonlat
+
+
+class MissingHeightError(ValueError):
+    """Raised when heights do not overlap with area to be corrected."""
+
+
+class IncompleteHeightWarning(UserWarning):
+    """Raised when heights only partially overlap with area to be corrected."""
+
+
+def forward_parallax(sat_lon, sat_lat, sat_alt, lon, lat, height):
+    """Calculate forward parallax effect.
+
+    Calculate the forward parallax effect.  When a satellite instrument
+    observes the Earth, the geolocation assumes it sees the Earth surface at
+    the geoid (elevation zero).  In reality, the ray may stop short of the
+    geoid as it observes, for example, a cloud or elevated ground.  This
+    function calculates the forward parallax effect.  If the view of a pixel at
+    location (lat, lon) is blocked by a cloud at height h, we calculate the
+    location of this blocking.
+
+    Calculate parallax correction based on satellite position and
+    (cloud top) height coordinates in geodetic (unprojected) coordinates.
+    This function calculates the latitude and longitude belonging to the
+    cloud top, based on the location of the satellite and the location
+    of the cloud.
+
+    For scenes that are only partly cloudy, the user might set the cloud top
+    height for clear-sky pixels to NaN.  This function will return a corrected
+    lat/lon as NaN as well.  The user can use the original lat/lon for those
+    pixels or use the higher level :class:`ParallaxCorrection` class.
+
+    This function assumes a spherical Earth.
+
+    Args:
+        sat_lon (number): Satellite longitude in geodetic coordinates [°]
+        sat_lat (number): Satellite latitude in geodetic coordinates [°]
+        sat_alt (number): Satellite altitude above the Earth surface [km]
+        lon (array or number): Longitudes of pixel or pixels to be corrected,
+            in geodetic coordinates [°]
+        lat (array or number): Latitudes of pixel/pixels to be corrected, in
+            geodetic coordinates [°]
+        height (array or number): Heights of pixels on which the correction
+            will be based.  Typically this is the cloud based height. [km]
+
+    Returns:
+        tuple[float, float]: New geolocation
+            New geolocation ``(lon, lat)`` for the longitude and
+            latitude that were to be corrected, in geodetic coordinates. [°]
+    """
+    X_sat = np.hstack(lonlat2xyz(sat_lon, sat_lat)) * sat_alt
+    X = np.stack(lonlat2xyz(lon, lat), axis=-1) * EARTH_RADIUS
```

Comment:
pep8 recommends small case for variable names, right?

Reply:
PEP-8 recommends ALLCAPS for constants: https://www.python.org/dev/peps/pep-0008/#constants

The Earth radius in this module is an import from `pyorbital.orbital`:

https://github.com/pytroll/pyorbital/blob/1bf73c834befb6d7f060c636dae85bb91230a556/pyorbital/orbital.py#L76

Reply:
Oh, it's about the `X`.  I hadn't even noticed this.  Will edit.